compiler/types.nim

```diff
@@ -1056,8 +1056,13 @@ proc commonSuperclass*(a, b: PType): PType =
     x = x.sons[0]
   var y = b
   while y != nil:
+    var t = y # bug #7818, save type before skip
     y = skipTypes(y, skipPtrs)
-    if ancestors.contains(y.id): return y
+    if ancestors.contains(y.id):
+      # bug #7818, special case with tyRef and tyPtr
+      # skip it anyway, they don't play nice with tyGenericInst
+      if k in {tyRef, tyPtr}: t = t.skipTypes(skipPtrs)
```

Comment:
I move it here to make the code intention clear.
skipTypes in line 1064 is basically a deferred  skipTypes from line 1060. 
it is a special case with tyRef/tyPtr similar with skipTypes inside proc commonType line 157,158 of sem.nim